src/Adapter/Http/ApachePassword.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\Authentication\Adapter\Http;
+
+use function base64_encode;
+use function chr;
+use function count;
+use function explode;
+use function hash_equals;
+use function mb_strlen;
+use function mb_substr;
+use function md5;
+use function min;
+use function pack;
+use function password_verify;
+use function preg_match;
+use function sha1;
+use function str_starts_with;
+use function strpos;
+use function strrev;
+use function strtr;
+
+/** @internal */
+final class ApachePassword
+{
+    private const BASE64  = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+    private const ALPHA64 = './0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
+
+    /**
+     * Verify if a clear-text password is correct given a hash value
+     *
+     * The username and realm are required in order to check credentials in digest format.
+     */
+    public static function verify(string $password, string $hash, string|null $username, string|null $realm): bool
+    {
+        if (str_starts_with($hash, '{SHA}')) {
+            $hash2 = '{SHA}' . base64_encode(sha1($password, true));
+
+            return hash_equals($hash, $hash2);
```

Comment:
I might be quite wrong but I feel like adding prefixes and b64 encode remove all the timing attack protection offered by `hash_equals()`.

I'd suggest doing the inverse and removing the prefix + b64 decode on `$hash`:

```suggestion
            return hash_equals(
                base64_decode(substr($hash, 5), true),
                sha1($password, true)
            );
```